### PR TITLE
gh-128911: Add tests on the PyImport C API

### PR DESCRIPTION
**Pull request by vstinner**
* Add Modules/_testlimitedcapi/import.c
* Add Lib/test/test_capi/test_import.py
* Remove _testcapi.check_pyimport_addmodule(): tests already covered by newly added tests.

* Issue: gh-128911

### CODE REVIEW

**Comment by vstinner** [comment]
cc @serhiy-storchaka @encukou

File: Modules/_testlimitedcapi/import.c [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+
+/* Test PyImport_GetModule() */
+static PyObject *
+pyimport_getmodule(PyObject *Py_UNUSED(module), PyObject *name)
+{
+    assert(!PyErr_Occurred());
+    PyObject *module = PyImport_GetModule(name);
+    if (module == NULL && !PyErr_Occurred()) {
+        Py_RETURN_NONE;
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: Nitpick: None has a special meaning in `sys.modules`, so it is better to use some other singleton, for example PyExc_KeyError.

File: Modules/_testlimitedcapi/import.c [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+{
+    return Py_XNewRef(PyImport_GetModuleDict());
+}
+
+
+/* Test PyImport_GetModule() */
+static PyObject *
+pyimport_getmodule(PyObject *Py_UNUSED(module), PyObject *name)
+{
+    assert(!PyErr_Occurred());
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c:41]: Add `NULLABLE(name)` to be able to test with NULL.

File: Modules/_testlimitedcapi/import.c [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    }
+    return module;
+}
+
+
+/* Test PyImport_AddModuleObject() */
+static PyObject *
+pyimport_addmoduleobject(PyObject *Py_UNUSED(module), PyObject *name)
+{
+    return Py_XNewRef(PyImport_AddModuleObject(name));
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c:56]: Add `NULLABLE(name)` to be able to test with NULL.

File: Modules/_testlimitedcapi/import.c [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+
+/* Test PyImport_AddModule() */
+static PyObject *
+pyimport_addmodule(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    const char *name;
+    if (!PyArg_ParseTuple(args, "s", &name)) {
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: Use `z#` to be able to test with non-UTF-8 string and with NULL.

File: Modules/_testlimitedcapi/import.c [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+
+/* Test PyImport_ImportModuleEx() */
+static PyObject *
+pyimport_importmoduleex(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    const char *name;
+    PyObject *globals, *locals, *fromlist;
+    if (!PyArg_ParseTuple(args, "sOOO",
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: Use `z#` and call `NULLABLE` after the parsing to be able to test with non-UTF-8 string and NULL. You could also use `O?`, but this test is purposed to be backported.

File: Modules/_testlimitedcapi/import.c [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    const char *name;
+    if (!PyArg_ParseTuple(args, "s", &name)) {
+        return NULL;
+    }
+
+    int res = PyImport_ImportFrozenModule(name);
+    if (res < 0) {
+        return NULL;
+    }
+    return PyLong_FromLong(res);
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: You can use `RETURN_INT()`.

File: Modules/_testlimitedcapi/import.c [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+/* Test PyImport_ExecCodeModuleWithPathnames() */
+static PyObject *
+pyimport_executecodemodulewithpathnames(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    const char *name;
+    PyObject *code;
+    const char *pathname;
+    const char *cpathname;
+    if (!PyArg_ParseTuple(args, "sOzz", &name, &code, &pathname, &cpathname)) {
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: Use `z#` instead of `s` and `z` to be able to test with non-UTF-8 strings.

File: Modules/_testlimitedcapi/import.c (hunk 2 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c:41]: It does crash.

File: Modules/_testlimitedcapi/import.c (hunk 1 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c]: Fixed

File: Modules/_testlimitedcapi/import.c (hunk 3 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c:56]: It does crash.

File: Modules/_testlimitedcapi/import.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_testlimitedcapi/import.c:41]: This is great too. We can check (manually) that it crashes.

But some functions can raise SystemError for NULL, and _this_ should have unit tests.

File: Modules/_testlimitedcapi/import.c (hunk 4 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c]: PyImport_AddModule(NULL) does crash (as written in the CRASHES comment).

**Comment by vstinner** [comment]
I will merge my PR as it is when the CI passed, to be able to add tests in my following https://github.com/python/cpython/pull/128912 PR.

I didn't add tests for non-UTF8 strings, nor NULL tests for `PyImport_ImportModuleLevel()`. For `PyImport_ImportModuleLevel()`, it's tricky because some arguments are ignored for `level=0` and I don't know how to write tests for `level > 0`.

@serhiy-storchaka: Maybe you would be interested to work on follow-up PR to increase the test coverage / test more cases?

File: Modules/_testlimitedcapi/import.c (hunk 2 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c:41]: `PyImport_Import(NULL)` does raise SystemError, I added a test for this case.

File: Modules/_testlimitedcapi/import.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/_testlimitedcapi/import.c]: Yes, but the wrapper does not crash, so we cannot test it from Python.

File: Modules/_testlimitedcapi/import.c [hunk 8]
```diff
@@ -94,7 +98,7 @@ static PyObject *
 pyimport_importmodule(PyObject *Py_UNUSED(module), PyObject *args)
 {
     const char *name;
-    if (!PyArg_ParseTuple(args, "s", &name)) {
+    if (!PyArg_ParseTuple(args, "z", &name)) {
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: `z#` to allow passing non-UTF-8 encoded bytes.

File: Modules/_testlimitedcapi/import.c [hunk 9]
```diff
@@ -39,9 +39,11 @@ static PyObject *
 pyimport_getmodule(PyObject *Py_UNUSED(module), PyObject *name)
 {
     assert(!PyErr_Occurred());
+    NULLABLE(name);
     PyObject *module = PyImport_GetModule(name);
     if (module == NULL && !PyErr_Occurred()) {
-        Py_RETURN_NONE;
+        PyErr_SetString(PyExc_KeyError, "PyImport_GetModule() returned NULL");
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/import.c]: I meant returning PyExc_KeyError as a value, not raising an exception. How can you differentiate it from PyExc_KeyError raised in PyImport_GetModule()?

**Review by serhiy-storchaka — COMMENTED** [review]
Do you allow me to commit directly in your branch? It is better to have a single clean commit in the logs than several commits that rewrite one other.

File: Modules/_testlimitedcapi/import.c (hunk 4 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c]: I modified the wrapper to use the `"z"` format to accept `NULL`.

File: Modules/_testlimitedcapi/import.c (hunk 9 above)
Comment (vstinner) [reply Modules/_testlimitedcapi/import.c]: Ah ok. I modified the code to return KeyError.

**Comment by vstinner** [comment]
@serhiy-storchaka: Ok, you can push changes to my branch.

**Comment by vstinner** [comment]
> I pushed my changes. Please review them.

I pushed a cleanup changes. With this cleanup, the change (additional tests) LGTM.

> It seems that I found bugs in the current code

Oh, interesting.

**Comment by vstinner** [comment]
@serhiy-storchaka: Thanks for additional tests and reviews.

I enabled auto-merge.
